ethcore/src/engines/clique/mod.rs

```diff
@@ -0,0 +1,766 @@
+		if is_checkpoint && signers == 0 {
+			Err(EngineError::CliqueCheckpointNoSigner)?
+		}
+
+		// Addresses must be be divisable by 20
+		if is_checkpoint && signers % ADDRESS_LENGTH != 0 {
+			Err(EngineError::CliqueCheckpointInvalidSigners(signers))?
+		}
+
+		// Ensure that the block doesn't contain any uncles which are meaningless in PoA
+		if *header.uncles_hash() != NULL_UNCLES_HASH {
+			Err(BlockError::InvalidUnclesHash(Mismatch {
+				expected: NULL_UNCLES_HASH,
+				found: *header.uncles_hash(),
+			}))?
+		}
+
+		// Ensure that the block's difficulty is meaningful (may not be correct at this point)
+		if *header.difficulty() != DIFF_INTURN && *header.difficulty() != DIFF_NOTURN {
+			Err(BlockError::DifficultyOutOfBounds(OutOfBounds {
+				min: Some(DIFF_NOTURN),
+				max: Some(DIFF_INTURN),
+				found: *header.difficulty(),
+			}))?
+		}
+
+		// All basic checks passed, continue to next phase
+		Ok(())
+	}
+
+	fn verify_block_unordered(&self, _header: &Header) -> Result<(), Error> {
+		// Nothing to check here.
+		Ok(())
+	}
+
+	/// Verify block family by looking up parent state (backfill if needed), then try to apply current header.
+	/// see https://github.com/ethereum/go-ethereum/blob/master/consensus/clique/clique.go#L338
+	fn verify_block_family(&self, header: &Header, parent: &Header) -> Result<(), Error> {
+		// Ignore genesis block.
+		if header.number() == 0 {
+			return Ok(());
+		}
+
+		// parent sanity check
+		if parent.hash() != *header.parent_hash() || header.number() != parent.number() + 1 {
+			Err(BlockError::UnknownParent(parent.hash()))?
+		}
+
+		// Ensure that the block's timestamp isn't too close to it's parent
+		let limit = parent.timestamp().saturating_add(self.period);
+		if limit > header.timestamp() {
+			let max = UNIX_EPOCH.checked_add(Duration::from_secs(header.timestamp()));
+			let found = UNIX_EPOCH.checked_add(Duration::from_secs(limit))
+				.ok_or(BlockError::TimestampOverflow)?;
+
+			Err(BlockError::InvalidTimestamp(OutOfBounds {
+				min: None,
+				max,
+				found,
+			}))?
+		}
+
+		// Retrieve the parent state
+		let parent_state = self.state(&parent)?;
+		// Try to apply current state, apply() will further check signer and recent signer.
+		let mut new_state = parent_state.clone();
+		new_state.apply(header, header.number() % self.epoch_length == 0)?;
+		new_state.calc_next_timestamp(header.timestamp(), self.period)?;
+		self.block_state_by_hash.write().insert(header.hash(), new_state);
+
+		Ok(())
+	}
+
+	fn genesis_epoch_data(&self, header: &Header, _call: &Call) -> Result<Vec<u8>, String> {
+		let mut state = self.new_checkpoint_state(header).expect("Unable to parse genesis data.");
+		state.calc_next_timestamp(header.timestamp(), self.period).map_err(|e| format!("{}", e))?;
+		self.block_state_by_hash.write().insert(header.hash(), state);
+
+		// no proof.
+		Ok(Vec::new())
+	}
+
+	// Our task here is to set difficulty
+	// TODO:(niklasad1): Return `Result<(), Error>` here instead
```

Comment:
Do we want to do this before merging in the PR, or leave it for a later date?

Reply:
I didn't realize that it belongs to the `EthEngine trait` so I removed that TODO should be fixed in the trait if it should be fixed at all sorry!

Reply:
https://github.com/goerli/parity-goerli/pull/82/files